### ISSUE TITLE
Restart issue with background fields continues to exist 

### ISSUE THREAD

**Issue by PrometheusPi** [ComputationalRadiationPhysics/picongpu#2137]
This is a follow up to #2113:

Test with the TWTS/plane wave background field setup (based on `dev`) still showed disappearing particles and wrong field values when used with restarts. Thus a restart issue still exists.

------
**UPDATE/important to mention** (thanks @ax3l for pointing this out)
**This is (most likely) an issue with the field restart!** The disappearing particles is the most obvious result, but is probably only caused by the wrong field values after restart. 

**UPDATE No. 2:**
I forgot to cherry-pick #2112 `aee7784672da6c3ded6fe00f47be9fdcd4b97948` - I know updated the plots and **the error is about a factor 2** in the simple setup presented below.

-----

The following difference could be seen for the same simulation setup - fist with restart every 10kth iteration, the second without any restart.

![n_macroparticles](https://user-images.githubusercontent.com/5121158/28201342-df38312c-6870-11e7-86d5-fd857b4e1749.png)

![fieldenergy](https://user-images.githubusercontent.com/5121158/28201352-ee35fe2a-6870-11e7-9619-92fb6568eb21.png)

As mentioned by @psychocoderHPC in #2113 I will try to find an easy setup to reproduce this bug based on 0.3.0.

mentioning:
@BeyondEspresso 
@steindev

**Comment by ax3l** [comment]
@PrometheusPi checkpoint/restart with ADIOS or HDF5?

**Comment by PrometheusPi** [comment]
@ax3l Restart with HDF5.

**Comment by PrometheusPi** [comment]
@psychocoderHPC I could reproduce this error using `0.3.0` with a cherry picked `daf0d002704b66abf6192995e95f5f7cfc73a30e` (#2113):

Based on the default Laser Wakefield example, but without particles and without Laser I used the following background field:

```c++
namespace picongpu
{
    class FieldBackgroundE
    {
    public:
        /* Add this additional field for pushing particles */
        static constexpr bool InfluenceParticlePusher = true;

        /* We use this to calculate your SI input back to our unit system */
        PMACC_ALIGN(m_unitField, const float3_64);

        HDINLINE FieldBackgroundE( const float3_64 unitField ) : m_unitField(unitField)
        {}

        /** Specify your background field E(r,t) here
         *
         * \param cellIdx The total cell id counted from the start at t = 0
         * \param currentStep The current time step */
        HDINLINE float3_X
        operator()( const DataSpace<simDim>& cellIdx,
                    const uint32_t currentStep ) const
        {
            /* example: periodicity of 20 microns ( = 2.0e-5 m) */
            constexpr float_64 period_SI(20.0e-6);
            constexpr float_64 omega_SI = 2.0 * PI * 2.9979e8 / period_SI;
            /* calculate cells -> SI -> m to microns*/
            const float_64 y_SI = cellIdx.y() * SI::CELL_HEIGHT_SI;
            /* note: you can also transform the time step to seconds by
             *       multiplying with DELTA_T_SI */

            /* specify your E-Field in V/m and convert to PIConGPU units */
            const float_X sinArg = precisionCast<float_X>( y_SI / period_SI * 2.0 * PI - currentStep *SI::DELTA_T_SI * omega_SI);
            return float3_X(0.0, math::sin( sinArg ) / m_unitField[1], 0.0);
        }
    };

```

and equivalent for the magnetic field.

The is an obvious (but in this case small) difference in the field energy:
![fieldenergy_0 3 0](https://user-images.githubusercontent.com/5121158/28206017-bb474136-6884-11e7-86fa-29a10a99de02.png)

Due to the periodic nature of these spikes, I will re-run a simulation without moving window and see if the spikes disappear.

**Comment by PrometheusPi** [comment]
This bug persist even without moving window:
![fieldenergy_0 3 0_nomoving](https://user-images.githubusercontent.com/5121158/28209118-655843b2-6891-11e7-964b-f26b61841db4.png)

**Comment by PrometheusPi** [comment]
The **error does not occur for normal/standard laser** input (in this case plane wave - with periodic (transversal) boundary conditions and moving window) which is treated by the Yee solver. 

![fieldenergy_0 3 0_planewave](https://user-images.githubusercontent.com/5121158/28213404-3dba55d8-68a6-11e7-8395-0dc8fef4af0d.png)

**Comment by ax3l** [comment]
> I could reproduce this error using 0.3.0 with a cherry picked daf0d002704b66abf6192995e95f5f7cfc73a30e (#2113):

careful, #2113 is not part of the last release and might actually cause the problem

**Comment by PrometheusPi** [comment]
Without the cherry-picked commit from the bug fix #2113 - the **same** error occurs with restart. 
![fieldenergy_0 3 0_nofix](https://user-images.githubusercontent.com/5121158/28217018-f7304ca4-68b3-11e7-8285-0eab471bd9bc.png)

**Comment by ax3l** [comment]
but looks less severe, probably `0.3.0` just needs the `step-1u` in the `Sub()` functors to be fixed...

**Comment by psychocoderHPC** [comment]
I thing I know where the field coming from and will create a bugfix.

**Comment by psychocoderHPC** [comment]
@PrometheusPi can we close this issue

**Comment by PrometheusPi** [comment]
Yes, we can close this issue.

**Comment by psychocoderHPC** [comment]
solved with #2139
